async-coap-uri/src/escape/unescape_uri.rs

```diff
@@ -202,6 +199,97 @@ impl<'a> UnescapeUri<'a> {
     }
 }
 
+#[cfg(feature = "std")]
+impl<'a> TryInto<String> for UnescapeUri<'a> {
+    type Error = DecodingError;
+
+    fn try_into(mut self) -> Result<String, Self::Error> {
+        let mut buffer = String::with_capacity(self.size_hint().0);
+
+        while let Some(ch) = self.next() {
+            if let Some(e) = self.had_error {
+                return Err(DecodingError::new(e, self.index()));
+            }
+
+            buffer.push(ch);
+        }
+
+        buffer.shrink_to_fit();
+        Ok(buffer)
+    }
+}
+
+/// This is returned, when an error occured while decoding.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct DecodingError {
+    inner: UnescapeError,
+    /// The index at which the error occured.
+    pub index: usize,
+}
+
+#[cfg(feature = "std")]
+impl ::std::error::Error for DecodingError {}
+
+impl fmt::Display for DecodingError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.inner)
+    }
+}
+
+impl DecodingError {
+    pub(crate) fn new(inner: UnescapeError, index: usize) -> Self {
+        Self { inner, index }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) enum UnescapeError {
+    /// found unescaped ascii control char
+    UnescapedAsciiControl(char),
+    /// found an unescaped space (' ')
+    Space,
+    /// there is no char after the `%`
+    MissingChar,
+    /// the char following a `%` must be an ascii character (in the range from 0-9 and a-f or A-F)
+    InvalidEscape(char),
+    /// We don't allow escaped ascii control codes for security reasons.
+    AsciiControl(char),
+    InvalidUtf8 {
+        buf: [u8; 4],
+        len: usize,
+    },
+    InvalidByte(u8),
+    UnexpectedChar(char),
+}
+
+impl fmt::Display for UnescapeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::UnescapedAsciiControl(c) => {
+                write!(f, "unescaped ascii control character `{:?}`", c)
+            }
+            Self::Space => write!(f, "unescaped space ` `"),
+            Self::MissingChar => write!(f, "missing char after `%`"),
+            Self::InvalidEscape(c) => write!(
+                f,
+                "the char after `%` must be a valid hex character `{:?}`",
+                c
+            ),
+            Self::AsciiControl(c) => write!(
+                f,
+                "ascii control codes are not allowed for security reasons `{:?}`",
+                c
+            ),
+            Self::InvalidUtf8 { buf, len } => write!(f, "invalid utf8 {:?}", &buf[..*len]),
+            Self::InvalidByte(b) => write!(f, "not a valid utf8 character `{}`", b),
+            Self::UnexpectedChar(c) => write!(f, "unexpected char `{:?}`", c),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl ::std::error::Error for UnescapeError {}
+
 impl<'a> Iterator for UnescapeUri<'a> {
```

Comment:
Before merging could you please look through this implementation for mistakes or error messages that can be improved?
There were some parts in this function that I did not understand, and I think there are some errors returned that are wrong or could be improved :)